internal/audit/audit.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2024 Mia srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package audit
+
+import (
+	"slices"
+
+	"github.com/google/uuid"
+)
+
+const (
+	AuditAdditionalDataGrantedPermissionKey = "authorization.permission"
+	AuditAdditionalDataGrantedBindingKey    = "authorization.binding"
+	AuditAdditionalDataGrantedRoleKey       = "authorization.role"
+)
+
+var reservedLabelKeys = []string{
+	AuditAdditionalDataGrantedPermissionKey,
+	AuditAdditionalDataGrantedBindingKey,
+	AuditAdditionalDataGrantedRoleKey,
+}
+
+type Audit struct {
+	AggregationID string
+	Authorization AuthzInfo
+	Subject       SubjectInfo
+	Request       RequestInfo
+	Labels        map[string]any
+}
+
+type AuthzInfo struct {
+	Allowed    bool
+	PolicyName string
+}
+
+type RequestInfo struct {
+	Body interface{} `audit:"body,omitempty"`
+	Path string      `audit:"path,omitempty"`
+	Verb string      `audit:"verb,omitempty"`
+}
+
+type SubjectInfo struct {
+	ID     string   `audit:"id"`
+	Groups []string `audit:"groups"`
+}
+
+// authzInfoToPrint defines the internal structure for an audit record and must be used by remapping
+// public Audit interface.
+type authzInfoToPrint struct {
+	Allowed    bool   `audit:"allowed"`
+	PolicyName string `audit:"policyName"`
+	Permission string `audit:"permission"`
+	BindingID  string `audit:"binding"`
+	RoleID     string `audit:"roleId"`
+}
+
+type auditToPrint struct {
+	ID            string           `audit:"id"`
+	AggregationID string           `audit:"aggregationId"`
+	Authorization authzInfoToPrint `audit:"authorization"`
+	Subject       SubjectInfo      `audit:"subject"`
+	Request       RequestInfo      `audit:"request"`
+	Labels        map[string]any   `audit:"labels"`
+}
+
+func (a *Audit) toPrint() auditToPrint {
+	return auditToPrint{
+		ID:            generateID(),
```

Comment:
>In this way, if a trace is done in both request and response (so 2 different traces), there are 2 different audit id? 

Yes since they are aggregated via the aggregationId

>For the log to stdout, it is useful the id?

I've put it up so that we don't have to think about this later when an id may be required for a different audit sink type.

For the audit log it maybe not be that useful, however I'd consider sharing a precise log (while navigating them with a tool such as grafana). I could send the precise query targeting a specific trace and I find this valuable enough to keep this id